pkg/kernel/factory.go

```diff
@@ -30,21 +33,25 @@ func NewFactory(ctx context.Context, config cfg.Config, logger log.Logger, bp *b
 		stages:      make(stages),
 	}
 
-	if err := factory.build(); err != nil {
+	if factory.kernel, err = newKernel(ctx, config, logger); err != nil {
+		return nil, fmt.Errorf("can not create kernel: %w", err)
+	}
+
+	if err = factory.build(); err != nil {
 		return nil, fmt.Errorf("can not build kernel factory: %w", err)
 	}
 
 	return factory, nil
 }
 
-func (f *factory) GetKernel() Kernel {
-	k := New(f.ctx, f.config, f.logger, f.middlewares, f.stages)
+func (f *factory) GetKernel() (Kernel, error) {
+	f.kernel.init(f.middlewares, f.stages)
```

Comment:
Why do you need to store the kernel instance on the factory? And should it be secured as a proper singleton?